pkg/k8s/watchers/cilium_network_policy.go

```diff
@@ -155,36 +248,54 @@ func (k *K8sWatcher) onUpsertCNP(
 		return nil
 	}
 
+	// check if this cnp was referencing or is now referencing a
+	// CiliumCIDRGroup and update the relevant metric accordingly.
+	if len(getCIDRGroupRefs(cnp)) > 0 {
+		cidrGroupPolicies[key] = struct{}{}
+	} else {
+		delete(cidrGroupPolicies, key)
+	}
+	metrics.CIDRGroupPolicies.Set(float64(len(cidrGroupPolicies)))
+
 	// We need to deepcopy this structure because we are writing
 	// fields.
 	// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
 	cnpCpy := cnp.DeepCopy()
 
+	translationStart := time.Now()
+	translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
+	metrics.CIDRGroupTranslationTimeStats.WithLabelValues(cnpCpy.Namespace, cnpCpy.Name).
+		Observe(time.Since(translationStart).Seconds())
+
 	var err error
 	if ok {
-		err = k.updateCiliumNetworkPolicyV2(cs, oldCNP, cnpCpy, initialRecvTime)
+		err = k.updateCiliumNetworkPolicyV2(cs, oldCNP, translatedCNP, initialRecvTime)
 	} else {
-		err = k.addCiliumNetworkPolicyV2(cs, cnpCpy, initialRecvTime)
+		err = k.addCiliumNetworkPolicyV2(cs, translatedCNP, initialRecvTime)
 	}
 	if err == nil {
-		cache[key] = cnpCpy
+		cnpCache[key] = cnpCpy
 	}
 
 	k.K8sEventProcessed(metricLabel, action, err == nil)
 
 	return err
 }
 
-func (k *K8sWatcher) onDeleteCNP(
+func (k *K8sWatcher) onDelete(
```

Comment:
comment, no action necessary: I found it a bit odd that the delete case didn't need to do any resolution of the CIDRGroupRefs internally in the way that the upsert case does. However, this is correct since any deletion ends up doing an `api.Rule` (~= `.Spec`) lookup in `policy.Repository` for every rule with the CNP's labels, and deletes those rules by label. Therefore, it doesn't matter what CIDRGroupRefs are present in the deleted CNP since the entire rules will be deleted down at the daemon policy layer. :+1: 

@christarazi and I also looked into similar cases for Upsert where one policy exists, then we update it with additional rules, or update it to delete rules. The logic is similar in that case, basically the `daemon.policyAdd()` function grabs the `policy.Mutex`, deletes all of the old rules by CNP labels, adds all the new rules, (calculates identity refcount impacts), and releases the policy Repository mutex. So there should be no lifecycle implications for those cases.